### PR TITLE
Two pane

### PR DESCRIPTION
**Pull request by abenari**
The two-pane layout is combining the index page with new / edit pages.
It is opening the form in the context of the index page using ajax.
It is client side only change, and uses unobtrusive js.

To activate the two-pane, simply add a css class 'table-two-pane' to the index table.

### CODE REVIEW

File: app/assets/javascripts/two-pane.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  }
+});
+
+$(document).on('submit','.two-pane-right', function() {
+  two_pane_submit();
+  return false;
+});
+
+$(document).on('click', ".form-actions a", function(e) {
+  if ($('.table-two-pane').length) {
```
Comment (ohadlevy) [inline app/assets/javascripts/two-pane.js]: does this make sense to extract to a function?

**Comment by lzap** [comment]
Pls hold on merging this until 1.3 is out.

Days of nice and easy cherry-picking are over I guess :-P

File: app/assets/javascripts/two-pane.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+function clear_errors(){
+  $('.error .help-inline').hide();
+  $('.error').removeClass('error');
+  $('.tab-error').removeClass('tab-error');
+  $('.alert-error').remove();
+}
+
+// when ajax call hit a session time out it should handle the redirect to login correctly.
+function handle_redirect(response){
+  var redirect = response.redirect || $("form[action$='/login']", response).attr('action');
```
Comment (domcleal) [inline app/assets/javascripts/two-pane.js:135]: [done] The check for /login might need to be more flexible, since people sometimes mount Foreman at /foreman, so it might be /foreman/login?

File: app/views/auth_source_ldaps/_form.html.erb [hunk 3]
```diff
@@ -2,23 +2,34 @@
 
 <%= form_for @auth_source_ldap do |f| %>
   <%= base_errors_for @auth_source_ldap %>
+  <ul class="nav nav-tabs" data-tabs="tabs">
+    <li class="active"><a href="#primary" data-toggle="tab"><%= _("LDAP server") %></a></li>
+    <li><a href="#account" data-toggle="tab"><%= _("Account") %></a></li>
+    <li><a href="#attributes" data-toggle="tab"><%= _("Attributes mapping") %></a></li>
```
Comment (domcleal) [inline app/views/auth_source_ldaps/_form.html.erb]: Nitpick, "Attribute mappings" reads a little better I think.

**Comment by domcleal** [comment]
I quite like this, but hadn't been expecting to.  Good job :)
1. The multi_checkboxes look a bit odd when it's in the short list format in the pane.  Here, the title appears on one line, then the select all icon on the next line, then followed by the options.  In the normal edit page it works a bit better as the label is to the left.
2. Could you add another way to collapse the pane, other than the Cancel buttons?  I found myself clicking on the item in the list again, or trying to find an "X" to close it.
3. You can still get to the full edit pages by opening in a new tab - should we add Edit links to the last column on the list views so people can still find it?

File: app/views/users/_form.html.erb [hunk 4]
```diff
@@ -30,11 +31,13 @@
...
-      <% unless @editing_self -%>
+    </div>
+    <% unless @editing_self -%>
+      <div class="tab-pane" id="roles">
         <%= checkbox_f f, :admin if User.current.can_change_admin_flag? %>
         <%= multiple_checkboxes f, :roles, @user, Role.givable.for_current_user, {:label => _('Roles')} %>
-      <% end %>
-    </div>
+      </div>
+    <% end %>
```
Comment (domcleal) [inline app/views/users/_form.html.erb]: Nitpick, there's an unless block below this could be part of.

**Comment by sodabrew** [comment]
Love it, fewer full page loads are a win!

File: app/assets/javascripts/two-pane.js (hunk 2 above)
Comment (ohadlevy) [inline app/assets/javascripts/two-pane.js:135]: $= means ends with (see http://api.jquery.com/attribute-ends-with-selector)

we could however make sure we add users/login to be on the safe side

Comment (domcleal) [inline app/assets/javascripts/two-pane.js:135]: Ah, thanks.

**Comment by ares** [comment]
I like the simplicity. Even it's two pane UI with tabs in right pane it still looks pretty simple which is essential I believe. Few comments:

Green button to create a resource help disappears when second pane is opened. They are placed above the table so I think it's not necessary to hide them. I may need help for the given page. Also the help is not specific to list action. I think even more important is to keep green button. I don't want to click Cancel just to create new record or Cancel and New just to reset the form.

Another thing to consider is pagination links and number of records being displayed. Here I'm on doubt. It wouldn't look nice because the second pane size but we loose this info.

Now when forms have buttons aligned to right user must navigate from left part to right part of a screen. Maybe it could be handy to have buttons aligned to the left.

When second pane is loading we can see a thin table on the left and loading logo. Then the box with solid border appears. Could we display the logo within the box immediately? So only the content of the box would reload after it's fetched.

Last thing, sometimes it can be hard to select item in left pane table. E.g. architectures page. It would be nice if you could click anywhere in <td> to select particular row. Also background-color during hover could change to indicate that the row is clickable.

So that's my 2 cents. I'm not sure why we need two panes but this is definitely nice and clean two pane solution :-) Please keep it simple and do not add much JS to it :-)

**Comment by abenari** [comment]
@ares @domcleal,  thanks so much for taking the time to review my UI changes.
The reason I wish to adopt the two-pane design is because I like the fact that it allows to keep the crud of simple objects in a single page. I think it is also helping to have a common UI concepts with the kattelo engine.
However I wanted to eliminate any distraction while in edit mode, this is why I hide the title action buttons (green and others) as well as the pagination and didn't have an extra close button at the top.
I have added the close button at the top in the last fix. I hop I managed to make it blend with the background so that it will not take the focus.
Having the submit and cancel buttons at the right seems to me as balancing the page that is already very much left aligned, I do agree that it is more efficient to have them on the left.

**Comment by ares** [comment]
I understand your motivation now and I agree. I checked last version and it looks even better. Now the border is displayed during loading. Also cross in corner is very useful and intuitive. Still one thing to consider (last comment got wrong formatting so placing once again)

Sometimes it can be hard to select item in left pane table. E.g. architectures page. It would be nice if you could click anywhere in `<td>` to select particular row. Also background-color during hover could change to indicate that the row is clickable. I know it would change the UI that you had when you saw the full table but still I'd find that useful.

Other my comments are cosmetic changes and probably very individual so you can ignore them :-)

**Comment by GregSutcliffe** [comment]
@abenari I'm with @domcleal - I wasn't expecting to like it, but I have to say, it's really, really good! I agree with not hiding the New/Help Buttons. Also on long pages the scrolling down to find Submit can be irritating - could we keep the scrollbar inside the edit frame and the Submit button outside the frame? Just cosmetic though, not massively important.

Overall, big :+1:, it's very cool!

**Comment by ohadlevy** [comment]
thanks @abenari merged as f60942c and 22a48ce for Foreman 1.4
